### PR TITLE
[MMS] Fix mms

### LINKED ISSUE #25260: ⚠️ [Wav2Vec2-MMS] `pipeline` and `from_pretrained` fail to load the Wav2Vec2 MMS checkpoints
### System Info

- `transformers` version: 4.31.0
- Platform: Linux-5.15.109+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.1
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.0.1+cu118 (False)
- Tensorflow version (GPU?): 2.12.0 (False)
- Flax version (CPU?/GPU?/TPU?): 0.7.0 (cpu)
- Jax version: 0.4.13
- JaxLib version: 0.4.13
- Using GPU in script?: `No`
- Using distributed or parallel set-up in script?: `No`

### Who can help?

@sanchit-gandhi @patrickvonplaten 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Put together a quick colab to run the model as mentioned in [our documentation](https://huggingface.co/docs/transformers/model_doc/mms#loading) - [colab notebook](https://github.com/Vaibhavs10/scratchpad/blob/main/wav2vec2_mms_repro.ipynb)

code snippets:

`Pipeline`

```python
from transformers import pipeline

model_id = "facebook/mms-1b-all"
target_lang = "fra"

pipe = pipeline(model=model_id, model_kwargs={"target_lang": target_lang, "ignore_mismatched_sizes": True})
``` 

Error (full traceback in the [colab notebook](https://github.com/Vaibhavs10/scratchpad/blob/main/wav2vec2_mms_repro.ipynb)):
```
RuntimeError: Error(s) in loading state_dict for Wav2Vec2ForCTC:
	size mismatch for lm_head.weight: copying a param with shape torch.Size([154, 1280]) from checkpoint, the shape in current model is torch.Size([314, 1280]).
	size mismatch for lm_head.bias: copying a param with shape torch.Size([154]) from checkpoint, the shape in current model is torch.Size([314]).
	You may consider adding `ignore_mismatched_sizes=True` in the model `from_pretrained` method.
```

`Processor` + `Model`

```python
from transformers import Wav2Vec2ForCTC, AutoProcessor

model_id = "facebook/mms-1b-all"
target_lang = "fra"

processor = AutoProcessor.from_pretrained(model_id, target_lang=target_lang)
model = Wav2Vec2ForCTC.from_pretrained(model_id, target_lang=target_lang, ignore_mismatched_sizes=True)
```

Error (full traceback in the [colab notebook](https://github.com/Vaibhavs10/scratchpad/blob/main/wav2vec2_mms_repro.ipynb)):

```
RuntimeError: Error(s) in loading state_dict for Wav2Vec2ForCTC:
	size mismatch for lm_head.weight: copying a param with shape torch.Size([154, 1280]) from checkpoint, the shape in current model is torch.Size([314, 1280]).
	size mismatch for lm_head.bias: copying a param with shape torch.Size([154]) from checkpoint, the shape in current model is torch.Size([314]).
	You may consider adding `ignore_mismatched_sizes=True` in the model `from_pretrained` method.
```

Similar issues reported by @xenova here: https://github.com/huggingface/transformers/issues/24223#issuecomment-1661174505

### Expected behavior

The expected behaviour would be that dispite the mismatch the model weights are loaded and the mismatch is rectified via `load_adapter` for pipeline (as mentioned here:https://github.com/huggingface/transformers/issues/24223#issuecomment-1595856093)

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Fixes #25260.

The problem is that the model state_dict is retrieved before the weights are tied which in the case of MMS/Wav2Vec2 means before the state dict is rewritten to the correct expected structure since MMS/Wav2Vec2 loads adapter weights when modeling_utils calls `tie_weights`.

I'm not 100% sure if the moving `model.tie_weights()` up here a couple of lines is ok, but it's necessary to fix MMS.
I'm pretty sure it's fine because `tie_weights` should not fundamentally change the state_dict architectures for models != MMS.

I'm not able to fully pinpoint the reason for how this bug came to be, but as stated in #25260 loading MMS 
worked on the PR and without having `accelerate` installed it also worked on the main.

There were a couple of PRs that touched similar logic around at the same time or a bit later/sooner which might have caused the issue.

- https://github.com/huggingface/transformers/pull/24200
- https://github.com/huggingface/transformers/pull/24505
- https://github.com/huggingface/transformers/pull/24310

I might have accidentally also not synced my PR branch with "main" before merging so that between starting to work on it and merging a different logic creeped in.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -3280,6 +3282,7 @@ def _find_mismatched_keys(
 
             if len(resolved_archive_file) > 1:
                 resolved_archive_file = logging.tqdm(resolved_archive_file, desc="Loading checkpoint shards")
+
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:3285]:
```suggestion
```

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -3223,6 +3224,7 @@ def _find_mismatched_keys(
             folder = os.path.sep.join(resolved_archive_file[0].split(os.path.sep)[:-1])
         else:
             folder = None
+
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:3227]:
```suggestion
```

File: tests/models/wav2vec2/test_modeling_wav2vec2.py [hunk 3]
```diff
@@ -1151,6 +1151,40 @@ def get_logits(model, input_features):
...
+            with torch.no_grad():
+                logits = model(
+                    input_values=batch["input_values"].to(torch_device),
+                    attention_mask=batch["attention_mask"].to(torch_device),
+                ).logits
+            return logits
+
+        input_features = [np.random.random(16_000 * s) for s in [1, 3, 2, 6]]
+
+        model = Wav2Vec2ForCTC.from_pretrained("hf-internal-testing/tiny-random-wav2vec2-adapter", target_lang="fr", ignore_mismatched_sizes=True)
```
Comment (patrickvonplaten) [inline tests/models/wav2vec2/test_modeling_wav2vec2.py:1177]: better test! The other tests didn't check for mismatched vocab sizes actually. This should solve it

**Review by sgugger — APPROVED** [review]
This should be okay. If it breaks any of the stuff I did for the tied weights loading, the CI will tell us  as I added some tests :-)

**Comment by patrickvonplaten** [comment]
@ydshieh ok to merge or should we run some more tests?

File: tests/models/wav2vec2/test_modeling_wav2vec2.py [hunk 4]
```diff
@@ -1151,6 +1151,42 @@ def get_logits(model, input_features):
 
         self.assertTrue(torch.allclose(logits, logits_2, atol=1e-3))
 
+    def test_load_and_new_target_lang_adapter(self):
```
Comment (ydshieh) [inline tests/models/wav2vec2/test_modeling_wav2vec2.py:1154]: might be good to add a short comment to explain a bit what this test does, as this is for a very special situation.

**Review by ydshieh — APPROVED** [review]
Good for me, but I always prefer a short comment  to explain what the tests does for such particular thing.

File: tests/models/wav2vec2/test_modeling_wav2vec2.py (hunk 4 above)
Comment (ydshieh) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py:1154]: even just a link to issue and this PR page is fine for me.

Comment (patrickvonplaten) [inline tests/models/wav2vec2/test_modeling_wav2vec2.py:1154]:
```suggestion
    # test that loading adapter weights with mismatched vocab sizes can be loaded
    def test_load_target_lang_with_mismatched_size(self):
```

**Pull request merged**